File: leetcode/1-100/45-jump-game-ii/bench/jump_game_ii.c
// LeetCode #45 bench mirror — C, the greedy range-expansion matcher (★).
//
// Mirrors bench/jump_game_ii.kara: one cursor with farthest/current_end/jumps scalars,
// collapsing the layered BFS into a single scan. Build a reachable array once, punch one slot
// per iteration, fold the jump count into a rolling checksum. Same workload + sink as every
// other mirror.

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

static int64_t jump(const int64_t *nums, int64_t n) {
    int64_t jumps = 0, current_end = 0, farthest = 0;
    for (int64_t i = 0; i < n - 1; i++) {
        if (i + nums[i] > farthest) {
            farthest = i + nums[i];
        }
        if (i == current_end) {
            jumps++;
            current_end = farthest;
        }
    }
    return jumps;
}

int main(void) {
    const int64_t total = 200000;
    const int64_t modulus = 1000000007;
    const int64_t n = 1000;

    int64_t *nums = (int64_t *)malloc((size_t)n * sizeof(int64_t));
    for (int64_t a = 0; a < n; a++) {
        nums[a] = 1 + (a % 4);
    }

    int64_t acc = 0;
    for (int64_t k = 0; k < total; k++) {
        nums[k % n] = 1 + (k % 9);
        int64_t ans = jump(nums, n);
        acc = (acc * 131 + ans) % modulus;
    }

    free(nums);
    printf("%lld\n", (long long)acc);
    return 0;
}

File: leetcode/1-100/45-jump-game-ii/bench/jump_game_ii.rs
// LeetCode #45 bench mirror — Rust, the greedy range-expansion matcher (★).
//
// Mirrors bench/jump_game_ii.kara exactly: one cursor with farthest/current_end/jumps scalars,
// collapsing the layered BFS into a single scan. Build a reachable array once, punch one slot
// per iteration, fold the jump count into a rolling checksum. Same workload + sink as every
// other mirror.

fn jump(nums: &[i64], n: i64) -> i64 {
    let mut jumps = 0i64;
    let mut current_end = 0i64;
    let mut farthest = 0i64;
    let mut i = 0i64;
    while i < n - 1 {
        if i + nums[i as usize] > farthest {
            farthest = i + nums[i as usize];
        }
        if i == current_end {
            jumps += 1;
            current_end = farthest;
        }
        i += 1;
    }
    jumps
}

fn main() {
    let total: i64 = 200000;
    let modulus: i64 = 1000000007;
    let n: i64 = 1000;

    let mut nums: Vec<i64> = (0..n).map(|a| 1 + (a % 4)).collect();

    let mut acc: i64 = 0;
    let mut k: i64 = 0;
    while k < total {
        nums[(k % n) as usize] = 1 + (k % 9);
        let ans = jump(&nums, n);
        acc = (acc * 131 + ans) % modulus;
        k += 1;
    }

    println!("{}", acc);
}
